src/epcpm/sunspectocsv.py

```diff
@@ -0,0 +1,643 @@
+
+        Args:
+            filter_: a subset of a Fields object to be used as a filter
+
+        Returns:
+            tuple: a tuple of values that have been filtered specified by field values in filter_
+        """
+        return tuple(
+            value for value, f in zip(attr.astuple(self), attr.astuple(filter_)) if f
+        )
+
+
+point_fields = Fields(
+    block_offset=data_point_fields.block_offset,
+    size=data_point_fields.size,
+    type=data_point_fields.type_uuid,
+    scale_factor=data_point_fields.factor_uuid,
+    units=data_point_fields.units,
+)
+
+bitfield_fields = Fields(
+    block_offset=bitfield_fields.block_offset,
+    size=bitfield_fields.size,
+    type=bitfield_fields.type_uuid,
+)
+
+
+def export(
+    path: pathlib.Path,
+    sunspec_model: epyqlib.attrsmodel.Model,
+    parameters_model: epyqlib.attrsmodel.Model,
+    column_filter: Fields = None,
+    skip_output: bool = False,
+) -> None:
+    """
+    Generate the SunSpec model data .csv file.
+
+    Args:
+        path: path and filename for .csv file
+        sunspec_model: SunSpec model
+        parameters_model: parameters model
+        column_filter: columns to be output to .csv file
+        skip_output: skip output of the generated files, previously used for skip_sunspec
+
+    Returns:
+
+    """
+    if skip_output:
+        return
+
+    if column_filter is None:
+        column_filter = attr_fill(Fields, True)
+
+    builder = epcpm.sunspectocsv.builders.wrap(
+        wrapped=sunspec_model.root,
+        parameter_uuid_finder=sunspec_model.node_from_uuid,
+        parameter_model=parameters_model,
+        column_filter=column_filter,
+    )
+
+    csv_data = builder.gen()
+
+    with open(path.with_suffix(".csv"), "w", newline="") as csv_file:
+        csv_writer = csv.writer(csv_file, quoting=csv.QUOTE_ALL)
+        for data_row in csv_data:
+            csv_writer.writerow(data_row)
+
+
+@builders(epcpm.sunspecmodel.Root)
+@attr.s
+class Root:
+    wrapped = attr.ib(type=epcpm.sunspecmodel.Root)
+    column_filter = attr.ib(type=Fields)
+    parameter_uuid_finder = attr.ib(default=None, type=typing.Callable)
+    parameter_model = attr.ib(default=None, type=epyqlib.attrsmodel.Model)
+    sort_models = attr.ib(default=False, type=bool)
+
+    def gen(self) -> typing.List[str]:
+        """
+        CSV generator for the SunSpec Root class.
+
+        Returns:
+            list: the rows of CSV data to be output
+        """
+        csv_data = []
+
+        if self.sort_models:
+            children = sorted(
+                self.wrapped.children,
+                key=lambda child: (
+                    0 if isinstance(child, epcpm.sunspecmodel.Model) else 1,
+                    getattr(child, "id", math.inf),
+                ),
+            )
+        else:
+            children = list(self.wrapped.children)
+
+        model_offset = 2  # account for starting 'SunS'
```

Comment:
account for 'SunS' or account for ID & Length?

Reply:
This is correct. The beginning of the SunSpec models is identified by the characters 'SunS' as a way to identify the beginning of the SunSpec address space. This causes an offset of 2 from the starting SunSpec address to the first model.